contracts/staking/AbstractValueDistributor.sol

```diff
@@ -0,0 +1,211 @@
+        );
+
+        _update(user_);
+
+        _totalShares -= amount_;
+        _userDistributions[user_].shares -= amount_;
+
+        _afterRemoveShares(user_, amount_);
+    }
+
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function _distributeValue(address user_, uint256 amount_) internal virtual {
+        _update(user_);
+
+        require(amount_ > 0, "ValueDistributor: amount has to be more than 0");
+        require(
+            amount_ <= _userDistributions[user_].owedValue,
+            "ValueDistributor: insufficient amount"
+        );
+
+        _userDistributions[user_].owedValue -= amount_;
+
+        _afterDistributeValue(user_, amount_);
+    }
+
+    /**
+     * @notice Hook function that is called after shares have been added to a user's distribution.
+     *
+     * This function can be used to perform any additional logic that is required,
+     * such as transferring tokens.
+     *
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares added.
+     */
+    function _afterAddShares(address user_, uint256 amount_) internal virtual {}
+
+    /**
+     * @notice Hook function that is called after shares have been removed from a user's distribution.
+     *
+     * This function can be used to perform any additional logic that is required,
+     * such as transferring tokens.
+     *
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares removed.
+     */
+    function _afterRemoveShares(address user_, uint256 amount_) internal virtual {}
+
+    /**
+     * @notice Hook function that is called after value has been distributed to a user.
+     *
+     * This function can be used to perform any additional logic that is required,
+     * such as transferring tokens.
+     *
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value distributed.
+     */
+    function _afterDistributeValue(address user_, uint256 amount_) internal virtual {}
+
+    /**
+     * @notice Updates the cumulative sum of tokens that have been distributed.
+     *
+     * This function should be called whenever user shares are modified or value distribution occurs.
+     *
+     * @param user_ The address of the user.
+     */
+    function _update(address user_) internal {
+        _cumulativeSum = _getFutureCumulativeSum(block.timestamp);
+        _updatedAt = block.timestamp;
+
+        if (user_ != address(0)) {
+            UserDistribution storage userDist = _userDistributions[user_];
+
+            userDist.owedValue +=
+                (userDist.shares * (_cumulativeSum - userDist.cumulativeSum)) /
+                PRECISION;
+            userDist.cumulativeSum = _cumulativeSum;
+        }
+    }
+
+    /**
+     * @notice Gets the value to be distributed for a given time period.
```

Comment:
Please add a note that users will usually be required to override this function to provide custom distribution mechanics